nixops/backends/hetzner.py

```diff
@@ -51,16 +51,38 @@ def get_type(cls):
 
     def __init__(self, xml, config):
         MachineDefinition.__init__(self, xml, config)
-        x = xml.find("attrs/attr[@name='hetzner']/attrs")
-        assert x is not None
-        attrs = [("main_ipv4", "mainIPv4", "string"),
-                 ("create_sub_account", "createSubAccount", "bool"),
-                 ("robot_user", "robotUser", "string"),
-                 ("robot_pass", "robotPass", "string"),
-                 ("partitions", "partitions", "string")]
-        for var, name, valtype in attrs:
-            node = x.find("attr[@name='" + name + "']/" + valtype)
-            setattr(self, var, xml_expr_to_python(node))
+
+        self.main_ipv4 = config["hetzner"]["mainIPv4"]
+        assert type(self.main_ipv4) is str
```

Comment:
Use descriptors instead? Possibly http://traitlets.readthedocs.io.

Reply:
These are nice ideas, but here I'm simply unifying the style to use the same as the other backends (such as the EC2 one, https://github.com/NixOS/nixops/blob/f10999a111fba80e1dc597b7e14069ea620832af/nixops/backends/ec2.py#L39).

If we want to push for a change in style in general, we should probably tackle that separately for all backends and in coordination with whoever takes care of the those.

Regarding traitlets, I'm all for more type checking, but as far as I can tell it's an extra dependency so I can't tell whether or not we want to add it to nixops.

Reply:
Apart from that, those types should already be checked by the NixOS module system, so I think most of those asserts aren't really needed.